### PR TITLE
[proposal] Importer 

### PR DESCRIPTION
**Pull request by karsasmus**
Thanks for contributing to Postleaf! 💙💚
### Pull Request Summary

This is a work-in-progess proposal for the wordpress/ghost importer.
### Have you...
- [x] Checked that your code adheres to our code guidelines?

---
### Contributor License Agreement

We want Postleaf to remain free and open for everyone to use forever, so please don’t submit any code or content that is licensed or copyrighted without first getting written consent.

Before we accept your pull request, we ask that you agree to the following terms to ensure there are no conflicts with Postleaf’s license now or in the future.
1. I certify that the code I am contributing to this project is completely original or free of any known license or copyright.
2. I agree to allow the developer of Postleaf, A Beautiful Site, LLC, license or relicense my work at their discretion. By submitting this pull request, I agree to forfeit all copyrights, intellectual property rights, and royalties pertaining to my contribution.
3. By entering my name in the space provided below, I agree to all of the terms mentioned above.

Signed: [Marc Apfelbaum]

Date: [2016-08-26]

### CODE REVIEW

**Comment by claviska** [comment]
Just a heads up — I won't be able to get to this until after the weekend. Not ignoring it, just a bit busy this weekend! 😎

**Comment by karsasmus** [comment]
That's fine. 😄 
I'm looking forward to your response.

File: app/source/classes/Importer/XmlParser.php [hunk 1]
```diff
@@ -21,6 +21,7 @@ public function parse(){
         if (extension_loaded('simplexml')){
             $this->parseWithSimpleXml();
         } else {
+        	// Do we need parsing via xml_parser_create?
```
Comment (claviska) [inline app/source/classes/Importer/XmlParser.php]: I don't think we need to, as SimpleXML is [enabled by default](http://php.net/manual/en/simplexml.installation.php) as of PHP 5.1.2.

**Comment by claviska** [comment]
I think we're getting ahead of ourselves a bit. Imports for other systems should be available as a  plugin, and we don't have a plugin API to develop with yet.

What's going to be most useful _today_ is a script that parses WP exports into a simple format we can work with. For example:

```php
try {
    $data = parse_wp('path/to/wp-export.xml');
} catch(\Exception $e) {
    // Failed to parse...
}
```

After which `$data` will contain something we can easily iterate, like this:

```php
[
    'posts' => [
        [
            'slug' => 'post-slug',
            'title' => 'Post Title',
            'content' => '<p>Post content</p>',
            'pub_date' => '2016-01-02 03:04:05',
            'tags' => ['tag1', 'tag2', 'tag3']
            ...
        ],
        ...
    ],
    'tags' => [...],
    'users' => [...]
]
```

Getting the resulting data into Postleaf is the easy part. Getting it out of the export files is much harder. But once we have that nailed down (and the plugin API at least started), the rest will be cake. 🍰

Being a plugin, I think work on this should go into a separate repo. If we can get methods to parse WP and/or Ghost exports, I'll create an `importer` repo to track progress and isolate issues from the main project. 😎

**Comment by karsasmus** [comment]
This should now work as intended. 😄 
The function "parseFile" parses the xml and returns an array with posts, tags, users...

The parse_wp function could be implemented as follows:

```php
function parse_wp($file){
    $class = \Postleaf\Importer\ImportFactory::factory('wordpress', $file);
    return $class->parseFile();
}
```

I could add a wordpress xml/ghost json to the new repo, so we have data to test/import.

**Comment by karsasmus** [comment]
I imported my wordpress blog. 😃

**Comment by icaromh** [comment]
@karsasmus Can you give a quick explain how to use?

**Comment by karsasmus** [comment]
@icaromh it's really quick'n dirty for testing purpose.

1. Export your wordpress blog.
2. Upload this xml to content/upload 
(3. rename it to import.xml)

I added the code into index.php, after $app->run() (line 221)
```php
$app->run();

$x = \Postleaf\Importer\ImportFactory::factory('wordpress', '/home/ubuntu/workspace/app/content/uploads/import.xml');
$data = $x->parseFile();

foreach($data['user'] as $slug => $ds){
    try {
        echo 'Adding User ' . $slug . ' </br>';
        User::add($slug, $ds);
    } catch (\Exception $e){
        echo($e->getMessage());
        
    }
}

foreach($data['tags'] as $slug => $ds){
    try {
        echo 'Adding Tag ' . $slug . ' </br>';
        Tag::add($slug, $ds);
    } catch (\Exception $e){
        echo($e->getMessage());
    }
}

foreach($data['posts'] as $slug => $ds){
    try {
        echo 'Adding Post ' . $slug . ' </br>';
        Post::add($slug, $ds);
    } catch (\Exception $e){
        echo($e->getMessage());
    }
}
```
After that run you should see your posts, users and tags in postleaf.

But like Cory wrote, we do the import first without the classes.

**Comment by claviska** [comment]
I created a repo so we can separate the importer from the core. I'm going to start laying the foundation for plugins today so we can get this worked in sooner than later.

The importer repo is located here: https://github.com/Postleaf/importer

**Comment by icaromh** [comment]
Awesome!

I'll try run and give you guys the feedback

Em ter, 30 de ago de 2016 às 15:24, Cory LaViska <notifications@github.com>
escreveu:

**Comment by claviska** [comment]
> Do I understand you correctly, that we will refactoring the code for Version 1.1(?) if the plugin api is present?

Yes, I'll be adding the necessary events to support basic plugins soon. We'll use the importer as a test case.

We won't be waiting until 1.1, though. See https://github.com/Postleaf/postleaf/issues/39.

**Pull request closed without merging**